### PR TITLE
common: add VM-independent codec and text engines

### PR DESCRIPTION
**Pull request by youknowone**
## Summary

- add feature-gated Rust ports of the PyPy CJK codec state machines and preserve their mapping-table page/index layout
- add the runtime-independent binascii byte transforms shared by the Python adapters
- add strict and lenient IPv4/IPv6 text conversion plus its libc-parity corpus
- add WTF-8-preserving JSON string escaping and scanning
- expose slice/state APIs from rustpython-common without Python VM object dependencies

All four surfaces are opt-in. In particular, cjk-codecs keeps roughly 2.6 MB of generated mapping tables out of default rustpython-common consumers, while binascii only enables base64 and crc32fast for users that request it. This lets Pyre remove these engines from its frequently rebuilt interpreter/native workspace while making them reusable by either interpreter.

## Validation

- repository pre-commit hooks
- cargo fmt --all -- --check
- cargo clippy -p rustpython-common --features binascii,cjk-codecs,inet,json --all-targets -- -D warnings
- cargo test -p rustpython-common --features binascii,cjk-codecs,inet,json
- cargo test --workspace --exclude rustpython_wasm --exclude rustpython-venvlauncher --exclude rustpython-capi

## Summary by CodeRabbit

- **New Features**
  - Added broad CJK codec support, including Chinese, Japanese, Korean, Taiwanese, Hong Kong, and ISO-2022 encodings.
  - Added incremental encoder/decoder and stream reader/writer support.
  - Added shared utilities for binary conversions, JSON string handling, and IPv4/IPv6 parsing and formatting.

- **Improvements**
  - Improved validation and error handling for encoding, decoding, networking, and JSON operations.
  - Expanded test coverage for codec behavior, malformed input, round trips, and edge cases.

### CODE REVIEW

File: crates/common/src/binascii.rs [hunk 1]
```diff
@@ -0,0 +1,769 @@
...
+    let len = buffer.len();
+    let mut out_data = Vec::<u8>::with_capacity(len);
+    let mut idx = 0;
+
+    out_data.push(buffer[idx]);
+    idx += 1;
+
+    while idx < len {
+        if buffer[idx] == RUN_CHAR {
+            if buffer[idx + 1] == 0 {
```
[the remark this answers is not in the document]
Comment (youknowone) [reply crates/common/src/binascii.rs:670]: Fixed in #8623. rledecode_hqx now follows PyPy interp_hqx.rledecode_hqx: empty input returns empty bytes, a trailing RLE marker raises binascii.Incomplete, and an orphaned leading marker raises binascii.Error. Focused common-engine and public-adapter regressions are included.

File: crates/common/src/encodings/cjk/iso2022.rs [hunk 2]
```diff
@@ -0,0 +1,785 @@
...
+            let plane2 = matches!(mapping, Mapping::JisX0213_2000_2 | Mapping::JisX0213_2004_2);
+            if plane2 == (code & 0x8000 != 0) {
+                return Ok((code & 0x7fff, consumed));
+            }
+            None
+        }
+        Mapping::Gb2312 => {
+            raw_map_cn(&GBCOMMON_ENCMAP, &GBCOMMON_ENCMAP_DATA, c).filter(|code| code & 0x8000 == 0)
+        }
+        Mapping::Iso8859_1 | Mapping::Iso8859_7 => None,
```
[the remark this answers is not in the document]
Comment (youknowone) [reply crates/common/src/encodings/cjk/iso2022.rs:344]: I checked this against the upstream owner and am intentionally not applying it. PyPy _codecs_iso2022.c registers REGISTRY_ISO8859_1 and REGISTRY_ISO8859_7 with dummy_encoder; the default branch in ENCODER(iso2022) explicitly says G2/G3 encoding is unsupported and returns MBERR_INTERNAL. #8623 documents this decision and keeps the line-by-line PyPy shape.

File: crates/stdlib/src/cjkcodecs/multibytecodec.rs [hunk 3]
```diff
@@ -0,0 +1,1284 @@
...
+            }
+            Ok(())
+        }
+
+        #[pymethod]
+        fn reset(&self, vm: &VirtualMachine) -> PyResult<()> {
+            let mut local = self.inner.lock().clone();
+            let Some(pending) = local.pending.take() else {
+                return Ok(());
+            };
```
[the remark this answers is not in the document]
Comment (youknowone) [reply crates/stdlib/src/cjkcodecs/multibytecodec.rs:1236]: Fixed in #8623. MultibyteStreamWriter.reset now passes an empty string through the reset path when no text is pending, so ISO-2022 and HZ emit their reset sequences. The PR includes focused stream-writer coverage.

**Comment by youknowone** [comment]
Post-merge review follow-up is #8623. Correction after checking CPython 3.14: it removes the obsolete HQX APIs, applies the real CJK initial-state test fix, and documents why the reset and G2 suggestions are not compatible with the measured 3.14 behavior.

File: crates/common/src/binascii.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (youknowone) [reply crates/common/src/binascii.rs:670]: Correction after checking the Python 3.14 target: #8623 now removes rlecode_hqx and rledecode_hqx instead of hardening an API that CPython 3.14 no longer exposes. The common implementation, stdlib wrappers, and temporary tests are all gone.

File: crates/stdlib/src/cjkcodecs/multibytecodec.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (youknowone) [reply crates/stdlib/src/cjkcodecs/multibytecodec.rs:1236]: Correction after running the CPython 3.14.2 oracle: StreamWriter.reset writes no bytes immediately when no text is pending; the next write emits the required transition. #8623 therefore restores the existing behavior and does not apply this review suggestion.
